crates/polars-core/src/datatypes/any_value.rs

```diff
@@ -356,29 +358,52 @@ impl<'a> AnyValue<'a> {
             UInt64(_) => DataType::UInt64,
             Float32(_) => DataType::Float32,
             Float64(_) => DataType::Float64,
+            String(_) | StringOwned(_) => DataType::String,
+            Binary(_) | BinaryOwned(_) => DataType::Binary,
             #[cfg(feature = "dtype-date")]
             Date(_) => DataType::Date,
-            #[cfg(feature = "dtype-datetime")]
-            Datetime(_, tu, tz) => DataType::Datetime(tu, tz.clone()),
             #[cfg(feature = "dtype-time")]
             Time(_) => DataType::Time,
+            #[cfg(feature = "dtype-datetime")]
+            Datetime(_, tu, tz) => DataType::Datetime(*tu, (*tz).clone()),
             #[cfg(feature = "dtype-duration")]
-            Duration(_, tu) => DataType::Duration(tu),
-            Boolean(_) => DataType::Boolean,
-            String(_) => DataType::String,
+            Duration(_, tu) => DataType::Duration(*tu),
             #[cfg(feature = "dtype-categorical")]
-            Categorical(_, _, _) => DataType::Categorical(None, Default::default()),
+            Categorical(_, rev_map, arr) => {
+                if arr.is_null() {
```

Comment:
The `dtype` is very often used and should not be cloning a `RevMapping`. That's way too expensive. That's the reason we have the `Option`. I am not convinced going from `AnyValue` to back to dtype should be lossless (for performance reasons).

Reply:
I also found that part suspicious. All tests pass just fine without cloning the revmap, so I think we can just get rid of that piece of conversion logic. I'll push a new commit.

Except for `Categorical` and `Enum`, none of the other types seem problematic.

Reply:
It shouldn't cause problems. The code that creates the `AnyValue`s internally is responsible for keeping the `DataType` around.